File: dotnet/FlowerBI.Engine/QueryGeneration/Ordering.cs
using System;
using System.Collections.Generic;
using System.Linq;
using FlowerBI.Engine.JsonModels;

namespace FlowerBI;

public class Ordering(LabelledColumn column, bool descending, int index)
{
    public bool Descending { get; } = descending;

    public LabelledColumn Column { get; } = column;

    public int Index { get; } = index;

    public int? SelectedIndex { get; }

    public int? AggregatedIndex { get; }

    public Ordering(IColumn column, bool descending, int index)
        : this(new LabelledColumn(null, column), descending, index) { }

    public string Direction => Descending ? "desc" : "asc";

    public Ordering()
        : this((IColumn)null, true, 0)
    {
        AggregatedIndex = 0;
    }

    public Ordering(
        OrderingJson json,
        Schema schema,
        int selects = 0,
        int values = 0,
        int calcs = 0
    )
        : this(
            json.Column == null ? null : schema.GetColumn(json.Column),
            json.Descending,
            json.Index == null || json.Type == null ? 0
                : json.Type == OrderingType.Select && json.Index < selects ? json.Index.Value
                : json.Type == OrderingType.Value && json.Index < values
                    ? json.Index.Value + selects
                : json.Type == OrderingType.Calculation && json.Index < calcs
                    ? json.Index.Value + selects + values
                : throw new ArgumentOutOfRangeException(
                    "json",
                    $"Ordering index {json.Index} is out of range in {json.Type}"
                )
        )
    {
        switch (json.Type)
        {
            case OrderingType.Select:
                if (json.Index < selects)
                    SelectedIndex = json.Index;
                break;
            case OrderingType.Value:
                if (json.Index < values)
                    AggregatedIndex = json.Index;
                break;
            case OrderingType.Calculation:
                if (json.Index < calcs)
                    AggregatedIndex = values + json.Index;
                break;
        }
    }

    public static IList<Ordering> Load(
        IEnumerable<OrderingJson> orderings,
        Schema schema,
        int selects = 0,
        int values = 0,
        int calcs = 0
    ) =>
        orderings?.Select(x => new Ordering(x, schema, selects, values, calcs)).ToList()
        ?? new List<Ordering>();
}

File: js/packages/@flowerbi/query-generation/src/Ordering.ts
import { OrderingJson, OrderingType, LabelledColumn, IColumn, Schema, FlowerBIException } from './types';

export class Ordering {
  public readonly Descending: boolean;
  public readonly Column: LabelledColumn | null;
  public readonly Index: number;
  public readonly SelectedIndex?: number;
  public readonly AggregatedIndex?: number;

  constructor(
    column: LabelledColumn | null,
    descending: boolean,
    index: number,
    selectedIndex?: number,
    aggregatedIndex?: number
  ) {
    this.Descending = descending;
    this.Column = column;
    this.Index = index;
    this.SelectedIndex = selectedIndex;
    this.AggregatedIndex = aggregatedIndex;
  }

  static fromIColumn(column: IColumn, descending: boolean, index: number): Ordering {
    return new Ordering({ joinLabel: null, value: column }, descending, index);
  }

  static defaultOrdering(): Ordering {
    return new Ordering(null, true, 0, undefined, 0);
  }

  static fromJson(
    json: OrderingJson,
    schema: Schema,
    selects = 0,
    values = 0,
    calcs = 0
  ): Ordering {
    const column = json.column ? schema.getColumn(json.column) : null;
    
    let index = 0;
    let selectedIndex: number | undefined;
    let aggregatedIndex: number | undefined;

    if (json.index !== undefined && json.type !== undefined) {
      switch (json.type) {
        case OrderingType.Select:
          if (json.index < selects) {
            selectedIndex = json.index;
            index = json.index;
          } else {
            throw new FlowerBIException(`Ordering index ${json.index} is out of range in ${json.type}`);
          }
          break;
        case OrderingType.Value:
          if (json.index < values) {
            aggregatedIndex = json.index;
            index = json.index + selects;
          } else {
            throw new FlowerBIException(`Ordering index ${json.index} is out of range in ${json.type}`);
          }
          break;
        case OrderingType.Calculation:
          if (json.index < calcs) {
            aggregatedIndex = values + json.index;
            index = json.index + selects + values;
          } else {
            throw new FlowerBIException(`Ordering index ${json.index} is out of range in ${json.type}`);
          }
          break;
      }
    }

    return new Ordering(column, json.descending, index, selectedIndex, aggregatedIndex);
  }

  static Load(
    orderings: OrderingJson[] | undefined,
    schema: Schema,
    selects = 0,
    values = 0,
    calcs = 0
  ): Ordering[] {
    return orderings?.map(x => Ordering.fromJson(x, schema, selects, values, calcs)) ?? [];
  }

  get Direction(): string {
    return this.Descending ? 'desc' : 'asc';
  }
}